### ISSUE TITLE
amdclang requires ENABLE_OFFLOAD_CLANG_DEBUG_O3=OFF

### ISSUE THREAD

**Issue by jakurzak** [QMCPACK/qmcpack#5127]
Right now, with amdclang, `-g` requires `-O0`.
Is it still a problem for the upstream Clang?
I understand that the original motivation for `ENABLE_OFFLOAD_CLANG_DEBUG_O3=ON` were problems with the upstream Clang.
Is this still an issue or we could potentially change the default setting to `ENABLE_OFFLOAD_CLANG_DEBUG_O3=OFF`?

**Comment by ye-luo** [comment]
I think it should be default OFF and requires users to opt-in. Otherwise it catches users off guard.
Probably also remove the compiler filter to make it a generic option. @PDoakORNL thoughts?

**Comment by prckent** [comment]
Also noting/confirming that I was unable to build the offload version with amdclang with a cmake debug build (at least a month back). Would be good to fix the defaults. It is only developers that are going to be making this flavor of build.

**Comment by ye-luo** [comment]
@jakurzak could you elaborate `-g requires -O0`. Doesn't compile, doesn' run or just cannot debug kernels?

**Comment by jakurzak** [comment]
I'm hitting linking problems and the compiler team replied with "only -O0 supported with -g".

**Comment by PDoakORNL** [comment]
So I think I'm agreeing with @ye-luo here. It is not expected to have anything other than -O0 for a Debug build.  I would expect if I want to debug with compiler optimization in place that would require me to specify more detailed flags. However in many environments recently I've noticed that the CPU and GPU builds were decoupled such that I now always check if a particular package actually build the GPU code in way that matches the overall CMAKE_BUILD_TYPE, it seems not uncommon to discover that without specific flags the optimized release kernels are always built. I don't think we should emulate this.

Historically, using a debugger, the suggested strategy is always to start with -O0 and only attempt with optimization if reproducing a problem isn't possible. I don't think it is surprising that the compiler team doesn't care much if -g works with optimizations on.

My experience linking hip kernels to debug with rocgdb is that it was very touchy and that actually getting it working to allow stepping into kernels required a particular agreement between runtime amd libraries, kernel driver, and last time I had it working a hand built rocgdb from the github repo. There was a situation where  I could link but would have runtime errors if I tried to debug a kernel as well.

In summary:
if CMAKE_BUILD_TYPE=Debug everything should be -O0 and the offload target and hip code should also be -g -O0 and whatever else they suggest to have rocgdb work.

There still needs to be someway for developers to build optimized release kernels to run with Debug if they are not chasing a bug in the accelerated code.  I think we can do this already by directly manipulating different advanced link and compile flags in ccmake or on the command.
